Mathlib/Combinatorics/SimpleGraph/Matching.lean

```diff
@@ -309,4 +317,79 @@ lemma exists_maximal_isMatchingFree [Finite V] (h : G.IsMatchingFree) :
   obtain ⟨Gmax, hGmax⟩ := Finite.exists_le_maximal h
   exact ⟨Gmax, ⟨hGmax.1, ⟨hGmax.2.prop, fun _ h' ↦ hGmax.2.not_prop_of_gt h'⟩⟩⟩
 
+/-- A graph `G` consists of a set of cycles, if each vertex is either isolated or connected to
+exactly two vertices. This is used to create new matchings by taking the `symmDiff` with cycles.
+The definition of `symmDiff` that makes sense is the one for `SimpleGraph`. This is why this
+definition is for `SimpleGraph`, rather than `SimpleGraph.Subgraph`.
+-/
+def IsCycles (G : SimpleGraph V) := ∀ ⦃v⦄, (G.neighborSet v).Nonempty → (G.neighborSet v).ncard = 2
+
+/--
+Given a vertex with one edge in a graph of cycles this gives the other edge incident
+to the same vertex.
+-/
+lemma IsCycles.other_adj_of_adj (h : G.IsCycles) (hadj : G.Adj v w) :
+    ∃ w', w ≠ w' ∧ G.Adj v w' := by
+  simp_rw [← SimpleGraph.mem_neighborSet] at hadj ⊢
+  cases' (G.neighborSet v).eq_empty_or_nonempty with hl hr
+  · exact ((Set.eq_empty_iff_forall_not_mem.mp hl) _ hadj).elim
+  · have := h hr
+    obtain ⟨w', hww'⟩ := (G.neighborSet v).exists_ne_of_one_lt_ncard (by omega) w
+    exact ⟨w', ⟨hww'.2.symm, hww'.1⟩⟩
+
+open scoped symmDiff
+
+lemma Subgraph.IsPerfectMatching.symmDiff_spanningCoe_IsCycles
+    {M : Subgraph G} {M' : Subgraph G'} (hM : M.IsPerfectMatching)
+    (hM' : M'.IsPerfectMatching) : (M.spanningCoe ∆ M'.spanningCoe).IsCycles := by
+  intro v
+  obtain ⟨w, hw⟩ := hM.1 (hM.2 v)
+  obtain ⟨w', hw'⟩ := hM'.1 (hM'.2 v)
+  simp only [symmDiff_def, Set.ncard_eq_two, ne_eq, imp_iff_not_or, Set.not_nonempty_iff_eq_empty,
+    Set.eq_empty_iff_forall_not_mem, SimpleGraph.mem_neighborSet, SimpleGraph.sup_adj, sdiff_adj,
+    spanningCoe_adj, not_or, not_and, not_not]
+  by_cases hww' : w = w'
+  · simp_all [← imp_iff_not_or, hww']
+  · right
+    use w, w'
+    aesop
+
+/--
+A graph `G` is alternating with respect to some other graph `G'`, if exactly every other edge in
+`G` is in `G'`. Note that the degree of each vertex needs to be at most 2 for this to be
+possible. This property is used to create new matchings using `symmDiff`. The definition of
+`symmDiff` for `SimpleGraph` is the one that makes sense, so this property is defined
+for `SimpleGraph` as well.
+-/
+def IsAlternating (G  G' : SimpleGraph V) :=
+  ∀ ⦃v w w': V⦄, w ≠ w' → G.Adj v w → G.Adj v w' → (G'.Adj v w ↔ ¬ G'.Adj v w')
+
+lemma IsPerfectMatching.symmDiff_spanningCoe_of_isAlternating {M : Subgraph G}
+    (hM : M.IsPerfectMatching) (hG' : G'.IsAlternating M.spanningCoe) (hG'cyc : G'.IsCycles)  :
+    ((M.spanningCoe ∆ G').toSubgraph (M.spanningCoe ∆ G')
```

Comment:
Seems much clearer to write
```suggestion
    (SimpleGraph.toSubgraph (M.spanningCoe ∆ G')
```
?

It's sort of weird to use dot notation to write an implicit argument explicitly. Perhaps this suggests the design of `toSubgraph` needs changing.